accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,259 @@
+
+We propose to solve this by **adapting saltboot to send FQDN of `next-server`**, that means server which provided initrd and kernel during PXE boot. This can be obtained by parsing DHCP response for `next-server` field. DHCP response is stored and available in dracut environment where saltboot is operating. Mandatory requirement in this case would be that **reverse lookup of IP address must be correctly configured** as `next-server` fields is IP address.
+
+Alternative is to use DNS service to lookup `tftp` CNAME, this is discussed in alternatives below.
+
+When the FQDN of `next-server` is not in the Uyuni database, such is the case when terminal is connected directly to the Uyuni, then `pxeserver_id` would be set to `null`.
+
+This is related to the [Hub Retail extension](https://github.com/uyuni-project/uyuni-rfc/pull/45) discussion about how should Retail architecture look like in Hub environment.
+
+## Storing data in database
+
+As data are provided within salt event, implementing new event handler `SaltbootPxeUpdateEvent` with accompanying `SaltbootPxeUpdateMessageAction` should be sufficient.
+
+Event handler should also schedule generation of new PXE entry.
+
+## Exposing PXE entries to users
+
+In current implementation, consumer of the `pxe_update` event data is `pxe/terminal_entry` state which is part of PXE formula.
+
+In case of keeping the final action in salt space, there is a need to provide pillar data to the state:
+
+  * by supplying directly to state when scheduling pxe generating action.
+  * by exposing database data to salt pillar system.
+
+Exposing PXE data is not hard requirement but may have some benefit like enabling user to manually apply `pxe/terminal_entry` or similar state to (re)generate pxe entries. This option can save resources by making UI/XML-RPC API changes non-essential.
+
+Database can be exposed to pillar system by external pillar mechanism connecting to the database using postgresql external pillar:
+
+```
+postgres:
+  db: susemanager
+  host: localhost
+  pass: spacewalk
+  port: 5432
+  user: spacewalk
+
+ext_pillar:
+  - postgres:
+      'retail_terminals':
+         query: "SELECT S.name, root_device, salt_device, boot_image, kernel_parameters,mac_address FROM
+                (SELECT * FROM suseRetailMinionInfo as RMI, rhnServer as S WHERE RMI.pxeserver_id = S.id AND S.name LIKE %s) AS I JOIN rhnServer AS S ON I.server_id = S.id"
+         as_list: False
+         depth: 1
+```
+
+or by writing pillar files for given PXE server using already existing Java methods provided by `MinionPillarFileManager`.
+
+
+Keeping data hidden in database has a positive side by not cluttering pillar space with rarely used data.
+To allow user to manually trigger PXE entries regeneration it is needed to develop UI or XMLRPC API endpoint, proposed below.
```

Comment:
Not sure I got it all right up to this point - I'll try to rephrase and you tell me if I'm wrong :wink: 

Currently, relevant data is contained in the `suse/manager/pxe_update` event and it is saved to appropriate plain files (`01-mac_address_dash_separated`, `01:mac_address_colon_separated.grub2.cfg`)  via a Salt state that reacts to it.

What we want to change the role of those plain files - they will still be needed but they will not be the canonical source of information (originally in the event) but just a copy, a "cache" of it. We want the event to processed by Java code and the event handler to save event data onto the DB, so that it becomes the canonical source. So far so good.

IIUC, those plain files will still be needed, and should be regegerated starting from data in the DB. Which opens the problem of how to transfer such data from Server to Branch/Proxy.

We have two choices: pillars or XMLRPC API.

Is all the above correct to this point?

Reply:
Then one problem I see with the XMLRPC route is authentication and authorization. What credentials should the Branch Server use to get the XMLRPC session key?

Additional question: would such XMLRPC endpoints make sense outside of this scope, for any other general use?

Reply:
There is some misunderstanding. I am not proposing to use XMLRPL API to transfer data from server to branch. I still expect salt state to take care of generating the pxe entry. XMLRPC API is there for users to be able to regenerate pxe entries if needed.

The data are being passed using pillars. But not by writing them to the pillar file and exposing, but by passing them to the state application directly (e.g. like calling `salt $branch state.apply pxe/terminal_entry boot_image='...' root='...' ...`. We can do this by using proper LocalCall from salt netapi.)

Reply:
> The data are being passed using pillars. But not by writing them to the pillar file and exposing, but by passing them to the state application directly

Thanks for clarifying, I indeed had a confused understanding.

In the hypothesis pillar data was saved in the database instead of files, would it be an option to just store as a pillar, thus avoiding the need of a new database table and set of APIs?

In other words, how would this RFC change if we had solved the problem of pillar storage already?

Reply:
If data are just in pillar files and not in database:
- need to add/adapt mechanism to maintain sync between system entries and pillars with pxe entries
- no need for xmlrpc api, only documentation how to manually trigger salt state or edit them
- no need for database

We would still need:
- java based event handler to write pillar files

The biggest drawback is the need to maintain coherence between system entries and pillar files which may not be completely straightforward as the pillar files are for branch servers, however tracing terminal system entries.